dotnet/src/SemanticKernel/Kernel.cs

```diff
@@ -208,9 +212,13 @@ public async Task<SKContext> RunAsync(ContextVariables variables, CancellationTo
                     continue;
                 }
 
-                context = await skFunction.InvokeAsync(context, cancellationToken: cancellationToken).ConfigureAwait(false);
+                functionResult = await skFunction.InvokeAsync(context, cancellationToken: cancellationToken).ConfigureAwait(false);
+
+                context = functionResult.Context;
+
+                allFunctionResults.Add(functionResult);
```

Comment:
I'm just wondering why we need the results of all function executions. Is this something that was requested, or is it just nice to have? Can we postpone introducing it until it's requested? It's always easier to add later than to remove.

Reply:
I'm trying to resolve the case of passing information like token usage to the end user. Previously it was part of SKContext. In this PR, initially I've added `ModelResults` to KernelResult. But then it was not clear, either to use ModelResults from last function in pipeline or combine all ModelResults from all functions together. Having ModelResults from last function won't work, because if user will receive information that 50 tokens were used for function execution, it will be only for last function, not for all of them. (I think it works in this way at the moment, not sure it's correct though.) So, returning ModelResults from all functions sounds more reasonable, but then, why we should return just one `Value` as last function result but all ModelResults? So, I decided to return all FunctionResults, which will contain value for each function result separately, as well as ModelResults separately. It should provide more visibility to the users how each function performed.

Reply:
Resolving based on private discussions.